plugins/gnupgplugin/gpgmebase.cpp

```diff
@@ -0,0 +1,761 @@
+                                                                  const QVariantMap &customParameters,
+                                                                  QByteArray *generatedIV)
+{
+    Q_UNUSED(algorithm);
+    Q_UNUSED(blockMode);
+    Q_UNUSED(keySize);
+    Q_UNUSED(customParameters);
+    Q_UNUSED(generatedIV);
+    return Result(Result::OperationNotSupportedError,
+                  QStringLiteral("The GnuPG plugin doesn't support generation of initialisation vector."));
+}
+
+Result Daemon::Plugins::GnuPGPlugin::generateKey(const Key &keyTemplate,
+                                                 const KeyPairGenerationParameters &kpgParams,
+                                                 const KeyDerivationParameters &skdfParams,
+                                                 const QVariantMap &customParameters,
+                                                 Key *key,
+                                                 const QString &home)
+{
+    Q_UNUSED(skdfParams);
+    Q_UNUSED(customParameters);
+
+    if (!kpgParams.isValid()) {
+        return Result(Result::CryptoPluginKeyGenerationError,
+                      QStringLiteral("invalid key generation parameters."));
+    }
+    if (kpgParams.keyPairType() != KeyPairGenerationParameters::KeyPairDsa
+        && kpgParams.keyPairType() != KeyPairGenerationParameters::KeyPairRsa
+        && kpgParams.keyPairType() != KeyPairGenerationParameters::KeyPairCustom) {
+        return Result(Result::CryptoPluginKeyGenerationError,
+                      QStringLiteral("unsupported key pair algorithm."));
+    }
+    QVariantMap::ConstIterator name = kpgParams.customParameters().constFind("name");
+    if (m_protocol == GPGME_PROTOCOL_OpenPGP
+        && name == kpgParams.customParameters().constEnd()) {
+        return Result(Result::CryptoPluginKeyGenerationError,
+                      QStringLiteral("missing name custom parameter."));
+    }
+    QVariantMap::ConstIterator email = kpgParams.customParameters().constFind("email");
+    if (email == kpgParams.customParameters().constEnd()) {
+        return Result(Result::CryptoPluginKeyGenerationError,
+                      QStringLiteral("missing email custom parameter."));
+    }
+
+    GPGmeContext ctx(m_protocol, home);
+    if (!ctx)
+        return Result(Result::CryptoPluginKeyGenerationError, ctx.error());
+
+    QString gnupgKeyParms = "<GnupgKeyParms format=\"internal\">\n";
+    if (kpgParams.keyPairType() == KeyPairGenerationParameters::KeyPairDsa) {
+        gnupgKeyParms += "Key-Type: DSA\n";
+        gnupgKeyParms += "Key-Length: 1024\n";
+        gnupgKeyParms += "Subkey-Type: DSA\n";
+        gnupgKeyParms += "Subkey-Length: 1024\n";
+    } else if (kpgParams.keyPairType() == KeyPairGenerationParameters::KeyPairRsa) {
+        gnupgKeyParms += "Key-Type: RSA\n";
+        gnupgKeyParms += QStringLiteral("Key-Length: %1\n").arg(static_cast<RsaKeyPairGenerationParameters>(kpgParams).modulusLength());
```

Comment:
I'm surprised the static cast works, actually... maybe better to simply in-place construct: RsaKeyPairGenerationParameters(kpgParams).modulusLength();

Reply:
Updated.